gateway/faux-mgs/src/main.rs

```diff
@@ -0,0 +1,198 @@
+        value_parser = level_from_str,
+        help = "Log level for MGS client",
+    )]
+    log_level: Level,
+
+    #[clap(long)]
+    sp: SocketAddrV6,
+
+    #[clap(long, short, default_value = "3000")]
+    timeout_millis: u64,
+
+    #[clap(subcommand)]
+    command: Commands,
+}
+
+fn level_from_str(s: &str) -> Result<Level> {
+    if let Ok(level) = s.parse() {
+        Ok(level)
+    } else {
+        bail!(format!("Invalid log level: {}", s))
+    }
+}
+
+#[derive(Subcommand, Debug)]
+enum Commands {
+    /// Ask SP on which port it receives messages from us.
+    Discover,
+
+    /// Ask SP for its current state.
+    State,
+
+    /// Attach to the SP's USART.
+    UsartAttach {
+        /// Put the local terminal in raw mode.
+        #[clap(long)]
+        raw: bool,
+    },
+}
+
+fn main() -> Result<()> {
+    let args = Args::parse();
+    let decorator = slog_term::TermDecorator::new().build();
+    let drain = slog_term::FullFormat::new(decorator)
+        .build()
+        .filter_level(args.log_level)
+        .fuse();
+    let drain = slog_async::Async::new(drain).build().fuse();
+    let log = Logger::root(drain, o!("component" => "faux-mgs"));
+
+    let socket = UdpSocket::bind("[::]:0")
+        .with_context(|| "failed to bind UDP socket")?;
+    socket
+        .set_read_timeout(Some(Duration::from_millis(args.timeout_millis)))
+        .with_context(|| "failed to set read timeout on UDP socket")?;
+
+    let request_kind = match args.command {
+        Commands::Discover => RequestKind::Discover,
+        Commands::State => RequestKind::SpState,
+        Commands::UsartAttach { raw } => {
+            return usart::run(log, socket, args.sp, raw);
+        }
+    };
+
+    let response = request_response(&log, &socket, args.sp, request_kind)?;
+    println!("{response:?}");
+
+    Ok(())
+}
+
+fn request_response(
+    log: &Logger,
+    socket: &UdpSocket,
+    addr: SocketAddrV6,
+    kind: RequestKind,
+) -> Result<Result<ResponseKind, ResponseError>> {
+    let request_id = send_request(log, socket, addr, kind)?;
+    loop {
+        let message = recv_sp_message(log, socket)?;
+        match message.kind {
+            SpMessageKind::Response { request_id: response_id, result } => {
+                if response_id != request_id {
+                    warn!(
+                        log, "ignoring unexpected response id";
+                        "response_id" => response_id,
+                    );
+                    continue;
+                }
+                return Ok(result);
+            }
+            SpMessageKind::SerialConsole(_) => {
+                debug!(log, "ignoring serial console packet from SP");
+                continue;
+            }
+        }
+    }
+}
+
+// On success, returns the request ID we sent.
+fn send_request(
+    log: &Logger,
+    socket: &UdpSocket,
+    addr: SocketAddrV6,
+    kind: RequestKind,
+) -> Result<u32> {
+    static REQUEST_ID: AtomicU32 = AtomicU32::new(1);
```

Comment:
Why is this atomic? For mutability? I assume you made it static so you wouldn't have to pass it as a parameter.

Reply:
Yes, just for simplicity - if this function sends all packets, this allows it to keep its own monotonically increasing counter, freeing the caller up from worrying about it.